rfcs/type-byte-array.md

```diff
@@ -0,0 +1,195 @@
+
+Floating-point numbers are stored in a format specified by IEEE 754.
+
+`buffer.readstring(b: buffer, offset: number, count: number): string`
+
+Used to read a string of length 'count' from the buffer at specified offset.
+
+`buffer.writestring(b: buffer, offset: number, value: string, count: number?): ()`
+
+Used to write data from a string into the buffer at specified offset.
+
+If an optional 'count' is specified, only 'count' bytes are taken from the string. 'count' cannot be larger than the string length.
+
+---
+
+All offsets start at 0 (not to be confused with indices that start at 1 in Luau tables).
+This choice is made for both performance reasons (no need to subtract 1) and for compatibility with data formats that often describe field positions using offsets.
+While there is a way to solve the performance problem using luajit trick where table array part is allocated from index 0, this would mean that data in the buffer has 1 extra byte and this complicates the bounds checking.
+
+Offsets and 'count' numbers are cast to an integer in an implementation-defined way.
+
+Read and write operations for relevant types are little endian as it is the most common use case, and conversion is often trivial to do manually.
+
+Additionally, unaligned offsets in all operations are valid and behave as expected.
+
+Unless otherwise specified, if a read or write operation would cause an access outside the data in the buffer, an error is thrown.
+
+### Public C API
+
+`void* lua_tobuffer(lua_State* L, int idx, size_t* len);`
+
+Used to fetch buffer data pointer and buffer size at specified location.
+
+If there is no buffer at the location, `NULL` is returned and `len` is not modified.
+
+`void* lua_newbuffer(lua_State* L, size_t l);`
+
+Pushes new buffer of size `l` onto the stack.
+
+`void* luaL_checkbuffer(lua_State* L, int narg, size_t* len);`
+
+Similar to `lua_tobuffer`, but throws a tag error if there is no buffer at specified location.
+
+`int luaopen_buffer(lua_State* L);`
+
+Registers the 'buffer' library. If `luaL_openlibs` is used, that includes the 'buffer' library.
+
+`LUA_BUFFERLIBNAME`
+
+Macro containing the 'buffer' library name.
+
+## Drawbacks
+
+This introduces 'buffer' as a class type in global typing context and adds new global 'buffer' table.
+While class type might intersect with user-defined 'buffer' type, such type redefinitions are already allowed in Luau, so this should not cause new type errors.
+Same goes for the global table, users can already override globals like 'string', so additional of a new global is backwards-compatible, but new table will not be accessible in such a case.
+
+This increases the complexity of the VM a little bit, since support for new tagged type is required in interpreter loop and GC.
+
+There is also a string buffer C API; by having functions talk about 'buffer' (like `luaL_extendbuffer`) and use `luaL_Buffer`, it might be a point of confusion for C API users.
+
+## Extensions
```

Comment:
This entire section should probably be rewritten as part of "Alternatives" because we are not adding cursor based APIs or resizing support at this time.

Reply:
Changed.